### PR TITLE
Feature: SNS Event FilterPolicyScope attribute

### PR DESCRIPTION
**Pull request by mustafa-sadiq**
### Issue #, if available
N/A

### Description of changes
Adds the FilterPolicyScope attribute to the SNS Subscription created for a SNS function event source.
This attribute is supported by cloudformation:
https://aws.amazon.com/blogs/compute/introducing-payload-based-message-filtering-for-amazon-sns/ 

If you try to deploy a function with event 

```
Events:
  NotificationTopic:
    Type: SNS
    Properties:
      Topic: !Ref MySNSTopic
      Region: region
      FilterPolicy:
        store:
        - example_corp
        event:
        - anything-but: order_cancelled
        customer_interests:
        - rugby
        - football
        - baseball
        price_usd:
        - numeric:
          - '>='
          - 100
      FilterPolicyScope: MessageAttributes
```

`sam deploy` gives this output

```
Error: Failed to create changeset for the stack: sam-hello-world, ex: Waiter ChangeSetCreateComplete failed: Waiter encountered a terminal failure state: For expression "Status" we matched expected path: "FAILED" Status: FAILED. Reason: Transform AWS::Serverless-2016-10-31 failed with: Invalid Serverless Application Specification document. Number of errors found: 1. Resource with id [HelloWorldFunctionNotificationTopic] is invalid. property FilterPolicyScope not defined for resource of type SNS
```

### Description of how you validated changes
I deployed a transformed template with the FilterPolicyScope: MessageAttributes added.

### Checklist

- [x] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [x] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [x] Using correct values
    - [x] Using wrong values
- [x] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Deploy a Function with a SNS event source and add property:
`FilterPolicyScope: MessageAttributes`

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/internal/schema_source/sam-docs.json
Comment (hoffa) [inline samtranslator/internal/schema_source/sam-docs.json:357]: I assume this was added manually? `sam-docs.json` is generated automatically from https://github.com/awsdocs/aws-sam-developer-guide once updated. Instead you should leave out the assignment in `aws_serverless_function.py` and remove this. See also https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#making-schema-changes

File: samtranslator/model/eventsources/push.py
Comment (hoffa) [inline samtranslator/model/eventsources/push.py:489]: Avoid potential bugs in the future (see https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines; it's a pass-through value and we don't control the underlying resource):

```suggestion
        "FilterPolicyScope": PassThroughProperty(False),
```

File: tests/translator/input/function_with_sns_event_source_all_parameters.yaml
Comment (hoffa) [inline tests/translator/input/function_with_sns_event_source_all_parameters.yaml:1]: What changed in this file?

File: samtranslator/model/eventsources/push.py
Comment (mustafa-sadiq) [reply samtranslator/model/eventsources/push.py]: added suggested change, thank you!

File: samtranslator/internal/schema_source/sam-docs.json
Comment (mustafa-sadiq) [reply samtranslator/internal/schema_source/sam-docs.json]: change reverted, thank you!

File: tests/translator/input/function_with_sns_event_source_all_parameters.yaml
Comment (mustafa-sadiq) [reply tests/translator/input/function_with_sns_event_source_all_parameters.yaml:1]: I think this is eol git change, I made this commit from gitpod. I'm not sure how to resolve it. Ill update if I figure out.

The only change in translator input tests is adding the following to the SNS Event properties:

`FilterPolicyScope: MessageAttributes`

File: samtranslator/internal/schema_source/sam-docs.json
Comment (hoffa) [reply samtranslator/internal/schema_source/sam-docs.json]: Sorry I probably wasn't clear, what I wanted to say that it should just be:

```python
    FilterPolicyScope: Optional[PassThroughProp]
```

Without the ` = snseventproperties("FilterPolicyScope")` part (the assignment only adds the documentation to the schema), and without the change to `sam-docs.json`. Once done, run `make schema` to update the schema (CI will complain otherwise).

File: tests/translator/input/function_with_sns_event_source_all_parameters.yaml
Comment (mustafa-sadiq) [reply tests/translator/input/function_with_sns_event_source_all_parameters.yaml]: fixed, I made the commit with git config change

`git config --global core.autocrlf false`

File: samtranslator/internal/schema_source/sam-docs.json
Comment (mustafa-sadiq) [reply samtranslator/internal/schema_source/sam-docs.json]: added the property back without assignment, all ran `make schema`. Thank you!

File: samtranslator/model/eventsources/push.py
Comment (mustafa-sadiq) [reply samtranslator/model/eventsources/push.py]: this change broke `make test`. I don't see PassThroughProperty being used elsewhere as well.

Comment (hoffa) [reply samtranslator/model/eventsources/push.py]: Ah, please import it on line 9:

```
from samtranslator.model import PropertyType, PassThroughProperty, ResourceMacro
```

File: tests/translator/input/function_with_sns_event_source_all_parameters.yaml
Comment (hoffa) [reply tests/translator/input/function_with_sns_event_source_all_parameters.yaml]: Nice! We should be able to fix it at the source by ensuring it uses `\n` for line endings.

File: samtranslator/model/eventsources/push.py
Comment (mustafa-sadiq) [reply samtranslator/model/eventsources/push.py]: imported. ran `make test` as well. thank you!

**Review by GavinZZ — COMMENTED** [review]
The changes look good. A question before I approval. I see that `FilterPolicyScope` supports two possible value, `MessageAttributes` or `MessageBody`. Is there any reason we aren't testing `MessageBody` as an input scope value?

**Comment by mustafa-sadiq** [comment]
> Thanks for opening the PR. The changes look good. I have one question before I approve. I see that [`FilterPolicyScope`](https://docs.aws.amazon.com/sns/latest/dg/sns-message-filtering.html) supports two possible value, `MessageAttributes` or `MessageBody`. Is there any reason we aren't testing `MessageBody` as an input?

The test will pass no matter what I change the value to. I set it to `InvalidValue` and it would pass through. I am not sure how to set an integration test so that CloudFormation will fail so I will leave that up to the team.

**Review by hoffa — APPROVED** [review]

**Review by GavinZZ — APPROVED** [review]
Thanks for the contribution!!

**Comment by mustafa-sadiq** [comment]
thank you @hoffa and @GavinZZ! I am new to open-source and appreciate the guidance and reviews! <3

**Pull request merged**
